python_modules/libraries/dagster-duckdb/dagster_duckdb/io_manager.py

```diff
@@ -138,16 +144,23 @@ def _get_cleanup_statement(table_slice: TableSlice) -> str:
     being written.
     """
     if table_slice.partition:
-        return (
-            f"DELETE FROM {table_slice.schema}.{table_slice.table}\n"
-            + _time_window_where_clause(table_slice.partition)
+        partition_where = (
+            _static_where_clause(table_slice.partition)
+            if isinstance(table_slice.partition.partition, str)
+            else _time_window_where_clause(table_slice.partition)
         )
+        return f"DELETE FROM {table_slice.schema}.{table_slice.table}\n" + partition_where
```

Comment:
Hi @jamiedemaria 
As I mentioned in Slack about the performance, when I was testing my implementation of static and multi-dimensional partitions, I notice a huge performance drop (from 5 seconds materialisation to 5 minutes and more). I tracked it down to this line and using `DELETE FROM` statement which apparently becomes significant when we deal with large number of rows. In our case, this is the case where we deal with static partitions. Cutting long story short, I could regain the performance by bypassing the direct `DELETE` through creating an new table with rows which we want to keep, dropping the original table, and then recreating the original table from the temporary table, and finally dropping the temporary table. My `_get_cleanup_statement` and `_get_where_clause` functions look like below:
```
def _get_cleanup_statement(table_slice: TableSlice) -> str:
    """
    Returns a SQL statement that deletes data in the given table to make way for the output data
    being written.
    """
    if table_slice.partition:
        where = _get_where_clause(table_slice.partition, delete=True)
        st0 = """DROP TABLE IF EXISTS temporary;"""
        st1 = f"""CREATE TABLE temporary as select * FROM {table_slice.schema}.{table_slice.table} {where};\n"""
        st2 = f"""DROP TABLE {table_slice.schema}.{table_slice.table};\n"""
        st3 = f"""CREATE TABLE {table_slice.schema}.{table_slice.table} AS SELECT * FROM temporary;\n"""
        st4 = """DROP TABLE temporary;"""
        return st0, st1, st2, st3, st4
        # return (
        #     f"DELETE FROM {table_slice.schema}.{table_slice.table}\n"
        #     + _get_where_clause(table_slice.partition)
        # )
    else:
        return f"DELETE FROM {table_slice.schema}.{table_slice.table}"

def _get_where_clause(table_partition: TablePartition, delete: bool = False) -> str:
    time_partitioned: bool = False
    static_partitioned: bool = False
    multi_partitioned: bool = False
    start_dt_str: str = None
    end_dt_str: str = None
    static_part_str: str = None
    DUCKDB_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
    log = get_dagster_logger()
    log.info(table_partition)

    if table_partition.time_window:
        start_dt, end_dt = table_partition.time_window
        start_dt_str = start_dt.strftime(DUCKDB_DATETIME_FORMAT)
        end_dt_str = end_dt.strftime(DUCKDB_DATETIME_FORMAT)
        time_window_key = table_partition.time_window_key
        time_partitioned = True

    if table_partition.static_part:
        static_part_str = table_partition.static_part_key
        static_part_curr_key = int(table_partition.static_part_curr_key)
        static_partitioned = True
    
    if static_partitioned and time_partitioned:
        multi_partitioned = True

    if time_partitioned:
        if delete:
            start = f"""WHERE {time_window_key} <= '{start_dt_str}' """
            end = f"""OR {time_window_key} > '{end_dt_str}'"""
            where_clause = start + end
        else:
            start = f"""WHERE {time_window_key} >= '{start_dt_str}' """
            end = f"""AND {time_window_key} < '{end_dt_str}'"""
            where_clause = start + end
    elif static_partitioned:
        if not delete:
            where_clause = f"""WHERE {static_part_str} != {static_part_curr_key}"""
        else:
            where_clause = f"""WHERE {static_part_str} = {static_part_curr_key}"""
    elif multi_partitioned:
        if delete:
            start_col = f"""WHERE {static_part_str} != {static_part_curr_key} """
            start = f"""({time_window_key} <= '{start_dt_str}' """
            end = f"""OR {time_window_key} > '{end_dt_str}') """
            where_clause = start_col + start + end
        else:
            start_col = f"""WHERE {static_part_str} = {static_part_curr_key} """
            start = f"""{time_window_key} >= '{start_dt_str}' """
            end = f"""AND {time_window_key} < '{end_dt_str}' """
            where_clause = start_col + start + end
    else:
        raise ValueError("I just ran out of options!")
    log.info(where_clause)
    return where_clause
```
I currently have a `delete` argument for testing purposes but this can be improved, if these make sense :)

Reply:
thank you for the detail - this is really good to know! I think for the sake of keeping this PR easy to review, i'll keep the delete statement as is and then look to incorporate this in a follow up PR